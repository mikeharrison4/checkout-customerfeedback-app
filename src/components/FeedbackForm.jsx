import React from 'react';
import { Formik, Form, Field } from 'formik';
import content from '../data/feedbackForm.json';
import FieldInput from './FieldInput';

const FeedbackForm = () => {
  return (
    <div className="text-white">
      <h1 className="text-5xl font-bold tracking-wider pb-12">
        {content.title}
      </h1>
      <Formik
        initialValues={{ name: '', email: '' }}
        validateOnBlur={false}
        validate={values => {
          const errors = {};
          if (!values.name) { // make sure to not allow any numbers here
            errors.name = 'Required';
          } else if (!/^[_A-z\s]*((-|\s)*[_A-z\s])*$/.test(values.name)) {
            errors.name = 'Invalid name';
          }
          if (!values.email) {
            errors.email = 'Required';
          } else if (!/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i.test(values.email)) {
            errors.email = 'Invalid email address';
          }
          return errors;
        }}
        onSubmit={(values, { setSubmitting }) => {
          setTimeout(() => {
            alert(JSON.stringify(values, null, 2));
            setSubmitting(false);
          }, 400);
        }}
      >
        {({ isSubmitting }) => (
          <Form className='feedback-form'>
            <div className="flex justify-between">
              <Field
                component={FieldInput}
                type="text"
                name="name"
                pattern="[a-z_A-Z\s]+"
                label="Name"
              />
              <Field
                component={FieldInput}
                name="email"
                type="email"
                label="Email"
              />
            </div>
            <div className="my-8">
              <button type="submit" disabled={isSubmitting}>
                Submit
              </button>
            </div>
          </Form>
        )}
      </Formik>
    </div>
  );
};

export default FeedbackForm;
