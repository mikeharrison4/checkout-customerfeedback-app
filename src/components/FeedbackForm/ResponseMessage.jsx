import React, { useEffect } from 'react';

const ResponseMessage = ({ message, isError }) => {
  const [showPopup, setShowPopup] = React.useState(true);

  React.useEffect(() => {
    setTimeout(() => {
      setShowPopup(false);
    }, 3000);
  }, []);

  if (showPopup) {
    return (
      <div className={`absolute top-0 right-0 m-3 text-white ${isError ? 'bg-pink-500' : 'bg-green-500'}`}>
        <div className="flex justify-between p-3">
          { message }
        </div>
      </div>
    );
  }

  return null;
};

export default ResponseMessage;
