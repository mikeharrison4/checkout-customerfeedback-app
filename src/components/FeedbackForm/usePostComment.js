import { useMutation, useQueryClient } from 'react-query';
import axios from 'axios';

const postComment = async (formValues) => await axios.post(
  'http://localhost:3001/comments',
  { ...formValues }
);

export const usePostComment = () => {
  const queryClient = useQueryClient();

  const { mutate: addComment, isSuccess, isError } = useMutation(postComment, {
    onSuccess: () => queryClient.invalidateQueries('comments'),
  });

  return { addComment, isSuccess, isError };
};