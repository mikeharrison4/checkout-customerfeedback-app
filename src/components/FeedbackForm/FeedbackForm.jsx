import React from 'react';
import { Formik, Form, Field } from 'formik';
import content from '../../data/feedbackForm.json';
import FieldTextInput from './Fields/FieldTextInput';
import { validateForm } from '../../utils/formValidation';
import FieldRatingInput from './Fields/FieldRatingInput';
import FieldTextArea from './Fields/FieldTextArea';
import LoadingSpinner from '../Util/LoadingSpinner';
import ResponseMessage from './ResponseMessage';
import Button from '../Util/Button';
import { usePostComment } from './usePostComment';

const initialFormValues = {
  name: '',
  email: '',
  rating: 3,
  comment: ''
};

const FeedbackForm = () => {
  let responseMessage = '';

  const { addComment, isSuccess, isError } = usePostComment();

  const handleFormSubmit = async (values, { resetForm }) => {
    await addComment(values);
    resetForm({ values: '' });
  };

  if (isError) {
    responseMessage = content.responseMessageFailure;
  }

  if (isSuccess) {
    responseMessage = content.responseMessageSuccess;
  }

  return (
    <>
      <h1 className="text-5xl text-white font-bold tracking-wider pb-4">
        {content.title}
      </h1>
      <Formik
        initialValues={initialFormValues}
        validate={(values) => validateForm(values)}
        onSubmit={handleFormSubmit}
      >
        {({ isSubmitting }) => (
          <Form className="text-white">
            <div className="mt-8 flex justify-between">
              <Field
                component={FieldTextInput}
                name="name"
                pattern="[a-z_A-Z\s]+"
                className="mr-2"
                label={content.name}
              />
              <Field
                component={FieldTextInput}
                name="email"
                type="email"
                className="ml-2"
                label={content.email}
              />
            </div>
            <div className="mt-12">
              <Field
                component={FieldTextArea}
                name="comment"
                label={content.comment}
              />
            </div>
            <div className="mt-12 mb-8">
              <Field
                component={FieldRatingInput}
                name="rating"
                label={content.rating}
              />
            </div>
            <div>
              <Button
                id='submit'
                type="submit"
                disabled={isSubmitting}
                className="text-primary bg-white"
                aria-label={content.submit}
              >
                <span className="flex items-center">
                  { isSubmitting && <LoadingSpinner className="mr-4" /> }
                  {content.submit}
                </span>
              </Button>
            </div>
          </Form>
        )}
      </Formik>
      { responseMessage && (
        <ResponseMessage
          message={responseMessage}
          isError={isError}
        />
      ) }
    </>
  );
};

export default FeedbackForm;
