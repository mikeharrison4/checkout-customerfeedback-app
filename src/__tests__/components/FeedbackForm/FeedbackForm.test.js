import React from 'react';
import { shallow } from 'enzyme';
import FeedbackForm from '../../../components/FeedbackForm/FeedbackForm';
import { usePostComment } from '../../../components/FeedbackForm/usePostComment';
import { Formik } from 'formik';
import content from '../../../data/feedbackForm.json';

jest.mock('../../../components/FeedbackForm/usePostComment', () => ({
  usePostComment: jest.fn(),
}));

const resetForm = jest.fn();

describe('FeedbackForm.jsx',  () => {
  beforeEach(() => {
    usePostComment.mockImplementation(() => ({
      addComment: jest.fn(),
      isSuccess: false,
      isError: false,
    }));
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render correctly when the form has not been submitted',  () => {
    const renderedComponent = shallow(<FeedbackForm />);
    expect(usePostComment).toHaveBeenCalledTimes(1);
    expect(renderedComponent).toMatchSnapshot();
  });

  it('should render correctly when form has been submitted and isSuccess is true', () => {
    usePostComment.mockImplementation(() => ({
      addComment: jest.fn(),
      isSuccess: true,
      isError: false,
    }));
    const renderedComponent = shallow(<FeedbackForm />);
    expect(usePostComment).toHaveBeenCalledTimes(1);
    expect(renderedComponent).toMatchSnapshot();
  });

  it('should render correctly when form has been submitted and isError is true', () => {
    usePostComment.mockImplementation(() => ({
      addComment: jest.fn(),
      isSuccess: false,
      isError: true,
    }));
    const renderedComponent = shallow(<FeedbackForm />);
    expect(usePostComment).toHaveBeenCalledTimes(1);
    expect(renderedComponent).toMatchSnapshot();
  });

  it('should call handleFormSubmit when submit btn is clicked', async () => {
    const addComment = jest.fn();
    const formValues = { values: 'mockValues' };
    usePostComment.mockImplementation(() => ({
      addComment,
    }));
    const renderedComponent = shallow(<FeedbackForm />);
    const handleFormSubmit = renderedComponent.find('Formik').props().onSubmit;
    await handleFormSubmit(formValues, { resetForm });

    expect(addComment).toHaveBeenNthCalledWith(1, formValues);
    expect(resetForm).toHaveBeenCalledTimes(1);
  });

  // it('should call validation and return empty object when form is valid', async () => {
  //   const renderedComponent = shallow(<FeedbackForm />);
  //   const validate = renderedComponent.find('Formik').props().validate;
  //   const errors = validate({
  //     name: 'mockName', email: 'mockEmail@mockaroo.com', comment: 'mockComment'
  //   });
  //   expect(errors).toEqual({});
  // });
  //
  // it('should call validation and return suitable errors when form is not filled out correctly', async () => {
  //   const renderedComponent = shallow(<FeedbackForm />);
  //   const validate = renderedComponent.find('Formik').props().validate;
  //   const errors = validate({});
  //   expect(errors).toEqual({ name: 'Required', email: 'Required', comment: 'Required' });
  // });
  //
  // it('should call validation and return suitable errors when form is not filled out correctly 1', async () => {
  //   const renderedComponent = shallow(<FeedbackForm />);
  //   const validate = renderedComponent.find('Formik').props().validate;
  //   const errors = validate({
  //     name: 'mockName', email: 'mockInvalidEmail', comment: 'mockComment'
  //   });
  //   expect(errors).toEqual({ email: content.invalidEmail });
  // });

  it('should render the formik children fields correctly when isSubmitting is false', () => {
    const renderedComponent = shallow(<FeedbackForm />);
    const children = renderedComponent
      .find(Formik)
      .renderProp('children')({ isSubmitting: false });
    expect(children).toMatchSnapshot();
  });

  it('should render the formik children fields correctly when isSubmitting is true', () => {
    const renderedComponent = shallow(<FeedbackForm />);
    const children = renderedComponent
      .find(Formik)
      .renderProp('children')({ isSubmitting: true });
    expect(children).toMatchSnapshot();
  });
});

describe('Form Validation', () => {
  it('should call validation and return empty object when form values are valid', async () => {
    const renderedComponent = shallow(<FeedbackForm />);
    const validate = renderedComponent.find('Formik').props().validate;
    const errors = validate({
      name: 'mockName',
      email: 'mockEmail@mockaroo.com',
      comment: 'mockComment'
    });
    expect(errors).toEqual({});
  });

  it('should call validation and return suitable errors when all required form values are missing', async () => {
    const renderedComponent = shallow(<FeedbackForm />);
    const validate = renderedComponent.find('Formik').props().validate;
    const errors = validate({});
    expect(errors).toEqual({
      name: content.required,
      email: content.required,
      comment: content.required
    });
  });

  it('should call validation and return suitable error when email is invalid', async () => {
    const renderedComponent = shallow(<FeedbackForm />);
    const validate = renderedComponent.find('Formik').props().validate;
    const errors = validate({
      name: 'mockName', email: 'mockInvalidEmail', comment: 'mockComment'
    });
    expect(errors).toEqual({ email: content.invalidEmail });
  });

  it('should call validation and return suitable error when name is invalid', async () => {
    const renderedComponent = shallow(<FeedbackForm />);
    const validate = renderedComponent.find('Formik').props().validate;
    const errors = validate({
      name: 'mockInvalidName12345', email: 'mockEmail@mockaroo.com', comment: 'mockComment'
    });
    expect(errors).toEqual({ name: content.invalidName });
  });
});