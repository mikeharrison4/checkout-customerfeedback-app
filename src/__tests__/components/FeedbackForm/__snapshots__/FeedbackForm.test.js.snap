// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedbackForm.jsx should render correctly when form has been submitted and isError is true 1`] = `
<Fragment>
  <h1
    className="text-5xl text-white font-bold tracking-wider pb-4"
  >
    Customer Feedback
  </h1>
  <Formik
    initialValues={
      Object {
        "comment": "",
        "email": "",
        "name": "",
        "rating": 3,
      }
    }
    onSubmit={[Function]}
    validate={[Function]}
  >
    <Component />
  </Formik>
  <ResponseMessage
    isError={true}
    message="Oh no! Something went wrong. Please try again"
  />
</Fragment>
`;

exports[`FeedbackForm.jsx should render correctly when form has been submitted and isSuccess is true 1`] = `
<Fragment>
  <h1
    className="text-5xl text-white font-bold tracking-wider pb-4"
  >
    Customer Feedback
  </h1>
  <Formik
    initialValues={
      Object {
        "comment": "",
        "email": "",
        "name": "",
        "rating": 3,
      }
    }
    onSubmit={[Function]}
    validate={[Function]}
  >
    <Component />
  </Formik>
  <ResponseMessage
    isError={false}
    message="Comment posted successfully! Thanks for your feedback :)"
  />
</Fragment>
`;

exports[`FeedbackForm.jsx should render correctly when the form has not been submitted 1`] = `
<Fragment>
  <h1
    className="text-5xl text-white font-bold tracking-wider pb-4"
  >
    Customer Feedback
  </h1>
  <Formik
    initialValues={
      Object {
        "comment": "",
        "email": "",
        "name": "",
        "rating": 3,
      }
    }
    onSubmit={[Function]}
    validate={[Function]}
  >
    <Component />
  </Formik>
</Fragment>
`;

exports[`FeedbackForm.jsx should render the formik children fields correctly when isSubmitting is false 1`] = `
<Form
  className="text-white"
>
  <div
    className="mt-8 flex justify-between"
  >
    <Field
      className="mr-2"
      component={[Function]}
      label="Name"
      name="name"
      pattern="[a-z_A-Z\\\\s]+"
    />
    <Field
      className="ml-2"
      component={[Function]}
      label="Email"
      name="email"
      type="email"
    />
  </div>
  <div
    className="mt-12"
  >
    <Field
      component={[Function]}
      label="Leave a comment here..."
      name="comment"
    />
  </div>
  <div
    className="mt-12 mb-8"
  >
    <Field
      component={[Function]}
      label="Rating"
      name="rating"
    />
  </div>
  <div>
    <Button
      aria-label="Submit"
      className="text-primary bg-white"
      disabled={false}
      id="submit"
      type="submit"
    >
      <span
        className="flex items-center"
      >
        Submit
      </span>
    </Button>
  </div>
</Form>
`;

exports[`FeedbackForm.jsx should render the formik children fields correctly when isSubmitting is true 1`] = `
<Form
  className="text-white"
>
  <div
    className="mt-8 flex justify-between"
  >
    <Field
      className="mr-2"
      component={[Function]}
      label="Name"
      name="name"
      pattern="[a-z_A-Z\\\\s]+"
    />
    <Field
      className="ml-2"
      component={[Function]}
      label="Email"
      name="email"
      type="email"
    />
  </div>
  <div
    className="mt-12"
  >
    <Field
      component={[Function]}
      label="Leave a comment here..."
      name="comment"
    />
  </div>
  <div
    className="mt-12 mb-8"
  >
    <Field
      component={[Function]}
      label="Rating"
      name="rating"
    />
  </div>
  <div>
    <Button
      aria-label="Submit"
      className="text-primary bg-white"
      disabled={true}
      id="submit"
      type="submit"
    >
      <span
        className="flex items-center"
      >
        <LoadingSpinner
          className="mr-4"
        />
        Submit
      </span>
    </Button>
  </div>
</Form>
`;
